# <center><font color='MediumSeaGreen'> Desafío guiado - Introducción al ambiente de trabajo </font><center>
    
   
<center><b> Bárbara Pérez </b><center>

____________________________________________________________

## <font color='MediumSeaGreen'>Requerimientos</font>

<b><font color='MediumSeaGreen'>1.</font> Hola Mundo (1 pto): </b>
    Utilice el método <span style='background:WhiteSmoke'><font color='MediumSeaGreen'>print</font></span> 
    para mostrar en pantalla la siguiente
línea:

 <i>Hola Mundo, esta es mi primera incursión en Python <\i>

In [1]:
print ('Hola Mundo, esta es mi primera incursión en Python')

Hola Mundo, esta es mi primera incursión en Python


 _________

<b><font color='MediumSeaGreen'>2.</font> Ingreso de variables (2 ptos): </b> Ingrese una serie de datos respecto a su información personal. Esta información debe estar ingresada en variables para su posterior uso.

Ingrese los siguientes datos:

- Su nombre.
- Edad.
- Una lista con sus 3 actividades favoritas.
- ¿Tiene mascotas?

In [2]:
name = 'Bárbara Pérez'
edad= 27
actividades_favoritas=['estudiar Data Science', 'ver series', 'practicar pole dance sport']
tiene_mascotas=True

_______________

<b><font color='MediumSeaGreen'>3.</font> Imprimir variables (2 ptos): </b> 

- Imprima cada variable creada: acceda al segundo elemento de su lista.

- Pida los tipos de datos asociados.

- Interpole los datos en un método print para generar una presentación sobre usted.



In [3]:
print(f'Nombre: {name}.')
print(f'Edad: {edad}.')
print(f'Actividad favorita: {actividades_favoritas[1]}.')
print(f'¿Tienes mascotas?: {tiene_mascotas}.')

Nombre: Bárbara Pérez.
Edad: 27.
Actividad favorita: ver series.
¿Tienes mascotas?: True.


In [4]:
print(f'Tipo de dato de la variable \033[1mNombre\033[0m:{type(name)}')
print(f'Tipo de dato de la variable \033[1mEdad\033[0m: {type(edad)}')
print(f'Tipo de dato de la variable \033[1mActividades favoritas\033[0m: {type(actividades_favoritas)} ')
print(f'Tipo de dato de la variable \033[1mMascotas\033[0m: {type(tiene_mascotas)} ')

Tipo de dato de la variable Nombre:<class 'str'>
Tipo de dato de la variable Edad: <class 'int'>
Tipo de dato de la variable Actividades favoritas: <class 'list'> 
Tipo de dato de la variable Mascotas: <class 'bool'> 


In [5]:
print(f"""Hola, mi nombre es \033[1m{name}\033[0m.
Tengo \033[1m{edad}\033[0m años de edad.
Mis actividades favoritas son: \033[1m{actividades_favoritas[0]}\033[0m, \033[1m{actividades_favoritas[1]}\033[0m, y \033[1m{actividades_favoritas[2]}\033[0m.
¿Tengo mascotas?: \033[1m{tiene_mascotas}\033[0m.""")

Hola, mi nombre es Bárbara Pérez.
Tengo 27 años de edad.
Mis actividades favoritas son: estudiar Data Science, ver series, y practicar pole dance sport.
¿Tengo mascotas?: True.


______________

<b><font color='MediumSeaGreen'>4.</font> Identificar códigos erróneos (2 ptos): </b> a continuación se presenta una serie de códigos que generan errores. Corríjalos y comente cuáles fueron las fallas.

#Error entregado: 

<span style='background:WhiteSmoke'>
print(<font color='darkred'>'Estaba la pájara pinta sentada en el verde limón</font>)
</span>

In [6]:
#Solución: Se coloca una comilla simple al final
print('Estaba la pájara pinta sentada en el verde limón')

Estaba la pájara pinta sentada en el verde limón


#Error entregado:

<span style='background:WhiteSmoke'>
print(<font color='darkred'>'Mi nombre es' name 'y tengo' edad, 'años'</font>)
</span>


In [7]:
#Solución: agregar comas entre el texto y las variables
print('Mi nombre es',name, 'y tengo',edad, 'años')

Mi nombre es Bárbara Pérez y tengo 27 años


#Error entregado:

<span style='background:gainsboro'>
<b>import</b> pandas <b>as</b> pd</span>

<span style='background:gainsboro'><b>import</b> numpy <b>as</b> np</span>

In [8]:
#Corre sin problema
import pandas as pd


In [9]:
#Corre sin problema
import numpy as np

#Error entregado:

<span style='background:gainsboro'>
<font color='maroon'>"Ornitorrinco"</font> + <font color='maroon'>45</font>
</span>


In [10]:
#Solución:
#Transformar el int a str
#Se concatenan los str
print("Ornitorrinco" + str(45))

Ornitorrinco45


_________

<br><b><font color='MediumSeaGreen'>5.</font> Utilizar pandas y numpy (3 ptos): </b>

Para ello, sigue los siguientes pasos.



<font color='MediumSeaGreen'>A.</font> Importa <span style='background:gainsboro'>pandas</span> y  <span style='background:gainsboro'>numpy</span>.

<font color='MediumSeaGreen'>B.</font> Posteriormente importa el archivo <span style='background:gainsboro'>flights.csv</span>.

<font color='MediumSeaGreen'>C.</font> Solicita las primeras y últimas 5 observaciones de la tabla de datos.

<font color='MediumSeaGreen'>D.</font> Utiliza <span style='background:gainsboro'>describe</span> para obtener las medidas univariadas de la columna year.

<font color='MediumSeaGreen'>E.</font> Utiliza <span style='background:gainsboro'>value_counts</span> para obtener la frecuencia de meses y años de la tabla.

<font color='MediumSeaGreen'>F.</font> Guarda en variables las primeras y últimas 15 observaciones de la tabla.

<font color='MediumSeaGreen'>G.</font> Solicita la media, mediana y desviación estándar de la cantidad de pasajeros utilizando los métodos de <span style='background:gainsboro'>numpy</span>. Replique el procedimiento para las últimas y primeras 15 observaciones. Comente brevemente los resultados


<b><font color='MediumSeaGreen'>A.</font> Importa <span style='background:gainsboro'>pandas</span> y  <span style='background:gainsboro'>numpy</span>.</b>

In [11]:
import pandas as pd
import numpy as np

<b><font color='MediumSeaGreen'>B.</font> Posteriormente importa el archivo <span style='background:gainsboro'>flights.csv</span>.</b>


In [12]:
df = pd.read_csv('flights.csv')

<b><font color='MediumSeaGreen'>C.</font> Solicita las primeras y últimas 5 observaciones de la tabla de datos.
</b>

In [13]:
df.head() 
#Automaticamente devuelve 5 valores
#También se puede usar df[:5]

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [14]:
df.tail() 
#Automaticamente devuelve 5 valores
#También se puede usar df[:-5]

year      month  passengers
139  1960     August         606
140  1960  September         508
141  1960    October         461
142  1960   November         390
143  1960   December         432

<b><font color='MediumSeaGreen'>D.</font> Utiliza <span style='background:gainsboro'>describe</span> para obtener las medidas univariadas de la columna year.
</b>

In [15]:
df.describe()['year']

count     144.000000
mean     1954.500000
std         3.464102
min      1949.000000
25%      1951.750000
50%      1954.500000
75%      1957.250000
max      1960.000000
Name: year, dtype: float64

<b><font color='MediumSeaGreen'>E.</font> Utiliza <span style='background:gainsboro'>value_counts</span> para obtener la frecuencia de meses y años de la tabla.</b>


In [16]:
df['year'].value_counts()

1949    12
1950    12
1951    12
1952    12
1953    12
1954    12
1955    12
1956    12
1957    12
1958    12
1959    12
1960    12
Name: year, dtype: int64

In [17]:
df['year'].value_counts('%')

1949    0.083333
1950    0.083333
1951    0.083333
1952    0.083333
1953    0.083333
1954    0.083333
1955    0.083333
1956    0.083333
1957    0.083333
1958    0.083333
1959    0.083333
1960    0.083333
Name: year, dtype: float64

In [18]:
df['month'].value_counts()

January      12
February     12
March        12
April        12
May          12
June         12
July         12
August       12
September    12
October      12
November     12
December     12
Name: month, dtype: int64

<b><font color='MediumSeaGreen'>F.</font> Guarda en variables las primeras y últimas 15 observaciones de la tabla.
</b>

In [19]:
primeros_15= df.head(15) 

In [20]:
print(primeros_15)

    year      month  passengers
0   1949    January         112
1   1949   February         118
2   1949      March         132
3   1949      April         129
4   1949        May         121
5   1949       June         135
6   1949       July         148
7   1949     August         148
8   1949  September         136
9   1949    October         119
10  1949   November         104
11  1949   December         118
12  1950    January         115
13  1950   February         126
14  1950      March         141


In [21]:
ultimas_15 = df.tail(15) 

In [22]:
print(ultimas_15)

     year      month  passengers
129  1959    October         407
130  1959   November         362
131  1959   December         405
132  1960    January         417
133  1960   February         391
134  1960      March         419
135  1960      April         461
136  1960        May         472
137  1960       June         535
138  1960       July         622
139  1960     August         606
140  1960  September         508
141  1960    October         461
142  1960   November         390
143  1960   December         432


<b><font color='MediumSeaGreen'>G.</font> Solicita la media, mediana y desviación estándar de la cantidad de pasajeros utilizando los métodos de <span style='background:gainsboro'>numpy</span>. Replique el procedimiento para las últimas y primeras 15 observaciones. Comente brevemente los resultados</b>


In [23]:
mean = np.mean(df['passengers'])
median = np.median(df['passengers'])
sd = np.std(df['passengers'])

In [24]:
print(f"""Pasajeros:
Media: {mean}
Mediana: {median}
Desviación estandar:{sd}""")

Pasajeros:
Media: 280.2986111111111
Mediana: 265.5
Desviación estandar:119.54904153277147


***Replique el procedimiento para las últimas y
primeras 15 observaciones. Comente brevemente los resultados***

In [25]:
#Media, Mediana y desviación estandar con método NUMPY
mean2_np = np.mean(primeros_15['passengers'])
median2_np = np.median(primeros_15['passengers'])
sd2_np = np.std(primeros_15['passengers'])

print(f"""Primeros 15 datos con NUMPY
Media: {mean2_np}
Mediana: {median2_np}
Desviación estandar:{sd2_np}""")

Primeros 15 datos con NUMPY
Media: 126.8
Mediana: 126.0
Desviación estandar:12.681745410891462


In [26]:
#Media, Mediana y desviación estandar con método PANADS
mean2_pd = primeros_15['passengers'].mean()
median2_pd = primeros_15['passengers'].median()
sd2_pd = primeros_15['passengers'].std()

print(f"""Primeros 15 datos con PANDAS
Media: {mean2_pd}
Mediana: {median2_pd}
Desviación estandar:{sd2_pd}""")

Primeros 15 datos con PANDAS
Media: 126.8
Mediana: 126.0
Desviación estandar:13.126853610606226


PRIMEROS 15 DATOS :

| Librerias | Media | Mediana | Desviación estandar|
| :---: | :---: | :---: | :---: |
| numpy | 126.8 | 126.0| 12.681745410891462 |
| pandas   | 126.8 | 126.0)| 13.126853610606226 |

In [27]:
#Media, Mediana y desviación estandar con método NUMPY
mean3_np = np.mean(ultimas_15['passengers'])
median3_np = np.median(ultimas_15['passengers'])
sd3_np = np.std(ultimas_15['passengers'])

print(f"""Ultimos 15 datos con NUMPY
Media: {mean3_np}
Mediana: {median3_np}
Desviación estandar:{sd3_np}""")

Ultimos 15 datos con NUMPY
Media: 459.2
Mediana: 432.0
Desviación estandar:75.29426715671431


In [28]:
#Media, Mediana y desviación estandar con método PANDAS
mean3_pd = ultimas_15['passengers'].mean()
median3_pd = ultimas_15['passengers'].median()
sd3_pd = ultimas_15['passengers'].std()

print(f"""Ultimos 15 datos con PANDAS
Media: {mean3_pd}
Mediana: {median3_pd}
Desviación estandar:{sd3_pd}""")

Ultimos 15 datos con PANDAS
Media: 459.2
Mediana: 432.0
Desviación estandar:77.93697087115606


ULTIMOS 15 DATOS :

| Librerias | Media | Mediana | Desviación estandar|
| :---: | :---: | :---: | :---: |
| numpy | 459.2 | 432.0| 75.29426715671431 |
| pandas   | 459.2 | 432.0 | 77.93697087115606 |

Los resultados esperados son **distintos** solo en el caso de la **desviación estandar** ya que:

**numpy:** sqrt ( 1 / n * sumatoria (media - observacion) ^ 2 )

**pandas:** sqrt ( 1 / (n-1) * sumatoria (media - observacion) ^ 2 )

pandas le resta 1 elemento para que no coincida con la media


_____